pkg/informant/cgroup.go

```diff
@@ -0,0 +1,228 @@
+	MaxUpscaleWaitMillis uint
+}
+
+// ReceivedUpscale notifies s.upscaleEventsRecvr
+//
+// Typically, (*AgentSet).ReceivedUpscale() is also called alongside this method.
+func (s *CgroupState) ReceivedUpscale() {
+	s.upscaleEventsSendr.Send()
+}
+
+// setMemoryHigh updates the cgroup's value of memory.high, according to the current total
+// system memory.
+//
+// This method MUST be called while holding s.updateLock, or without s.handleCgroupSignals()
+// running (or any other method that may call this).
+func (s *CgroupState) setMemoryHigh() error {
+	var newMemHigh uint64
+	mib := float64(1 << 20) // Size of 1 MiB = 2^20 = 1 << 20
+
+	totalSystemMem, err := getTotalSystemMemory()
+	if err != nil {
+		return fmt.Errorf("Error getting system memory: %w", err)
+	}
+
+	newMemHigh = s.config.calculateMemoryHighValue(totalSystemMem)
+
+	klog.Infof(
+		"Total system memory is %d bytes (%v MiB). Setting cgroup memory.high to %d bytes (%v MiB)",
+		totalSystemMem, float64(totalSystemMem)/mib, newMemHigh, float64(newMemHigh)/mib,
+	)
+
+	s.mgr.MemoryHighEvent.Consume()
+
+	if err := s.mgr.SetHighMem(newMemHigh); err != nil {
+		return fmt.Errorf("Error setting cgroup memory.high: %w", err)
+	}
+
+	klog.Infof("Successfully set cgroup memory.high")
+	return nil
+}
+
+// handleCgroupSignals is an internal function that handles "memory high" signals from the cgroup
+func (s *CgroupState) handleCgroupSignalsLoop(config CgroupConfig) {
+	// FIXME: the process might exit after freezing the cgroup but before thawing it. This isn't
+	// really feasible to fix here; the parent process should handle the "thaw if frozen" check on
+	// exit.
+	//
+	// FIXME: we should have "proper" error handling instead of just panicking. It's hard to
+	// determine what the correct behavior should be if a cgroup operation fails, though.
+	//
+	// FIXME: enforce a minimum wait period between requesting scale-ups (otherwise can be bad when
+	// memory usage is oscillating around memory.high but there aren't available resources)
+
+	for {
+		// Wait for a new signal
+		select {
+		case err := <-s.mgr.ErrCh:
+			panic(fmt.Errorf("Error listening for cgroup signals: %w", err))
+		case <-s.upscaleEventsRecvr.Recv():
+			klog.Infof("Received upscale event. Updating cgroup memory.high")
+			func() {
+				s.updateLock.Lock()
+				defer s.updateLock.Unlock()
+
+				if err := s.setMemoryHigh(); err != nil {
+					panic(fmt.Errorf("Error handling upscale signal: %w", err))
+				}
+			}()
+		case <-s.mgr.MemoryHighEvent.Recv():
+			// Double-check that there isn't currently an upscale event.
+			select {
+			case <-s.upscaleEventsRecvr.Recv():
+				klog.Warningf("Received memory high event, but also upscale event. Handling upscale and ignoring memory high")
+				func() {
+					s.updateLock.Lock()
+					defer s.updateLock.Unlock()
+
+					if err := s.setMemoryHigh(); err != nil {
+						panic(fmt.Errorf("Error handling upscale signal: %w", err))
+					}
```

Comment:
maybe make these methods? 

all of the panics worry me? I think if you expect that all of these errors are unrecoverable by their nature, I'd just panic in the functions that produce them rather than go through the effort of returning errors that are just going to convert them to panics? 

(unless these errors represent invariant violations, panic'ing this much seems to be indicative of some sort of other design flaw)